modules/core/02-client/legacy/v100/store.go

```diff
@@ -0,0 +1,141 @@
+package v100
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/cosmos/ibc-go/modules/core/02-client/types"
+	clienttypes "github.com/cosmos/ibc-go/modules/core/02-client/types"
+	host "github.com/cosmos/ibc-go/modules/core/24-host"
+	"github.com/cosmos/ibc-go/modules/core/exported"
+	smtypes "github.com/cosmos/ibc-go/modules/light-clients/06-solomachine/types"
+	ibctmtypes "github.com/cosmos/ibc-go/modules/light-clients/07-tendermint/types"
+)
+
+// MigrateStore performs in-place store migrations from SDK v0.40 of the IBC module to v1.0.0 of ibc-go.
+// The migration includes:
+//
+// - Migrating solo machine client states from v1 to v2 protobuf definition
+// - Pruning all solo machine consensus states
+// - Pruning expired tendermint consensus states
+func MigrateStore(ctx sdk.Context, storeKey sdk.StoreKey, cdc codec.BinaryCodec) (err error) {
+	store := ctx.KVStore(storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, host.KeyClientStorePrefix)
+
+	var clients []string
+
+	// collect all clients
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		keySplit := strings.Split(string(iterator.Key()), "/")
+		if keySplit[len(keySplit)-1] != host.KeyClientState {
+			continue
+		}
+
+		// key is clients/{clientid}/clientState
+		// Thus, keySplit[1] is clientID
+		clients = append(clients, keySplit[1])
+	}
+
+	for _, clientID := range clients {
+		clientType, _, err := types.ParseClientIdentifier(clientID)
+		if err != nil {
+			return err
+		}
+
+		clientPrefix := []byte(fmt.Sprintf("%s/%s/", host.KeyClientStorePrefix, clientID))
+		clientStore := prefix.NewStore(ctx.KVStore(storeKey), clientPrefix)
+
+		bz := clientStore.Get(host.ClientStateKey())
+		if bz == nil {
+			return clienttypes.ErrClientNotFound
+		}
+
+		switch clientType {
+		case exported.Solomachine:
+			any := &codectypes.Any{}
+			if err := cdc.Unmarshal(bz, any); err != nil {
+				return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
+			}
+
+			clientState := &ClientState{}
+			if err := cdc.Unmarshal(any.Value, clientState); err != nil {
+				return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
+			}
```

Comment:
This took me a while to figure out. Since I regenerate v1 solomachine into legacy, it doesn't have any implementation code and isn't a exported.ClientState, so you cannot use the usual methods for decoding. 

The work around is to decode to Any and then decode the Any.Value (by passing the any.Type lookup)